plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/CancellableRateLimitedFluxIterator.java

```diff
@@ -0,0 +1,230 @@
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+
+/**
+ * An iterator that allows to subscribe to a reactive publisher and request more elements
+ * in batches as the iterator is consumed so an slow consumer is not overwhelmed by a fast
+ * producer. Additionally it provides the ability to cancel the subscription before the entire
+ * flux is consumed, for these cases it possible to provide a cleaner function that would be
+ * invoked for all the elements that weren't consumed before the cancellation. (i.e. it's
+ * possible to free the memory allocated for a byte buffer).
+ */
+class CancellableRateLimitedFluxIterator<T> implements Subscriber<T>, Iterator<T> {
+    private static final Subscription CANCELLED_SUBSCRIPTION = new Subscription() {
+        @Override
+        public void request(long n) {
+            // no op
+        }
+
+        @Override
+        public void cancel() {
+            // no op
+        }
+    };
+
+    private final int elementsPerBatch;
+    private final Queue<T> queue;
+    private final Lock lock;
+    private final Condition condition;
+    private final Consumer<T> cleaner;
+    private final AtomicReference<Subscription> subscription = new AtomicReference<>();
+    private final Logger logger = LogManager.getLogger(CancellableRateLimitedFluxIterator.class);
+    private volatile Throwable error;
+    private volatile boolean done;
+    private int emittedElements;
+
+    /**
+     * Creates a new CancellableRateLimitedFluxIterator that would request to it's upstream publisher
+     * in batches as specified in {@code elementsPerBatch}. Additionally, it's possible to provide a
+     * function that would be invoked after cancellation for possibly outstanding elements that won't by
+     * consumed downstream but need to be cleaned in any case.
+     * @param elementsPerBatch the number of elements to request upstream
+     * @param cleaner the function that would be used to clean unused elements
+     */
+    CancellableRateLimitedFluxIterator(int elementsPerBatch, Consumer<T> cleaner) {
+        this.elementsPerBatch = elementsPerBatch;
+        this.queue = new ArrayDeque<>(elementsPerBatch);
+        this.lock = new ReentrantLock();
+        this.condition = lock.newCondition();
+        this.cleaner = cleaner;
+    }
+
+    @Override
+    public boolean hasNext() {
+        for (; ; ) {
+            boolean isDone = done;
+            boolean isQueueEmpty = queue.isEmpty();
```

Comment:
How is this thread-safe when `ArrayDeque` isn't thread safe? Aren't we putting stuff on this queue on the Netty event-loop threads but running these checks on the `SNAPSHOT` or whatever pool that is reading from the blocking stream?

Reply:
Since `done` is volatile, in theory the writes must be visible in the `SNAPSHOT` threads. But I agree that this might be tricky, would you prefer to use an `ArrayBlockingQueue`? To be honest I was on the fence with this solution, the `ArrayBlockingQueue` solution would be easier to understand.

Reply:
I went ahead and replaced the queue in https://github.com/elastic/elasticsearch/pull/65140/commits/2c3eefa1117b3371a21f42ea7902058f274fb746